crates/nargo/src/cli/gates_cmd.rs

```diff
@@ -1,55 +0,0 @@
-use std::collections::HashMap;
-
-use acvm::ProofSystemCompiler;
-use clap::ArgMatches;
-use std::path::Path;
-
-use crate::cli::compile_cmd::compile_circuit;
-use crate::errors::CliError;
-
-pub(crate) fn run(args: ArgMatches) -> Result<(), CliError> {
-    let args = args.subcommand_matches("gates").unwrap();
-    let show_ssa = args.is_present("show-ssa");
-    let allow_warnings = args.is_present("allow-warnings");
-    count_gates(show_ssa, allow_warnings)
-}
-
-pub fn count_gates(show_ssa: bool, allow_warnings: bool) -> Result<(), CliError> {
-    let curr_dir = std::env::current_dir().unwrap();
-    count_gates_with_path(curr_dir, show_ssa, allow_warnings)
-}
-
-pub fn count_gates_with_path<P: AsRef<Path>>(
-    program_dir: P,
-    show_ssa: bool,
-    allow_warnings: bool,
-) -> Result<(), CliError> {
-    let compiled_program = compile_circuit(program_dir.as_ref(), show_ssa, allow_warnings)?;
-    let gates = compiled_program.circuit.gates.clone();
-
-    // Store counts of each gate type into hashmap.
-    let mut gate_counts: HashMap<&str, u32> = HashMap::new();
-    for gate in gates.iter() {
-        *gate_counts.entry(gate.name()).or_default() += 1;
-    }
-
-    // Sort gates by name alphabetically for consistent display.
-    let mut sorted_gate_counts: Vec<(&str, u32)> = gate_counts.into_iter().collect();
-    sorted_gate_counts.sort_by(|(a, _), (b, _)| a.partial_cmp(b).unwrap());
-
-    println!("Gates successfully counted\n");
-
-    println!("Total gates: {}\n", gates.len());
-
-    println!("By type:");
-
-    for (gate_type, count) in sorted_gate_counts {
-        println!("{}: {}", gate_type, count);
-    }
-
-    let backend = crate::backends::ConcreteBackend;
-    let exact_circuit_size = backend.get_exact_circuit_size(compiled_program.circuit);
```

Comment:
I find this information to be quite useful when writing circuits. Would it be possible to keep this? Perhaps we could add it to the output of `nargo prove`? Or `nargo compile` instead as all we need is the compiled program and don't need a witness to determine the number of constraints.

Reply:
What is the difference between the exact circuit size and the gates count of the circuit?

Reply:
The exact circuit size is the size of the final constraint system generated by barretenberg. For example, the Pedersen OPCODE might be one gate in our ACIR, but it is many more constraints underneath. Here is the result of `nargo gates` on the `pedersen_check` test:

<img width="1373" alt="Screen Shot 2022-12-09 at 1 23 54 PM" src="https://user-images.githubusercontent.com/43554004/206767839-72fa88b7-02d8-45ca-bca6-a886ac3f4806.png">

Reply:
So it is a useful information but it is quite backend specific and noir is not going to use it anytime soon, I think we should remove it from noir for now and add a log in the aztec backend when we generate a proof to display the constraint size, since we need to compute it anyways.

Reply:
Note this is a feature of nargo (the tool which uses noir) and not noir the compiler. 

I think the gate proportion count is not that useful and can be removed. The exact circuit size is akin to cargo allowing you to print out the actual assembly that gets emitted from your rust program instead of just the llvm-ir using for example `cargo rustc -- --emit asm`

One usecase being that the user wants to modify their circuit so its below a certain power of two. We don't want the user to be touching the backend for this usecase.

I have two questions based off of the screenshot:

- Since it was not clear what the difference was between the two, maybe we should change it to say "Backend specific circuit size" ?
- Why does it say 20 gates generated and then 21?

Reply:
The total gates is the number of gates after "optimisation", this number depends on the backend you use.